gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -188,270 +195,168 @@ public Object call() throws IOException {
-      TreeMap<Long, ByteString> dataChunkMap = new TreeMap<>();
-      do {
-        responseObserver =
-            new InsertChunkResponseObserver(uploadId, writeOffset, objectHasher, dataChunkMap);
-        // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
-        stub.withDeadlineAfter(WRITE_STREAM_TIMEOUT.toMillis(), MILLISECONDS)
-            .insertObject(responseObserver);
-        try {
-          responseObserver.done.await();
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new IOException(
-              String.format("Resumable upload failed for '%s'", getResourceString()), e);
+      long writeOffset = retriesAttempted > 0 ? getCommittedWriteSize(uploadId) : 0;
+      objectHasher = retriesAttempted > 0 ? objectHasher : Hashing.crc32c().newHasher();
+
+      responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset);
+      // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
+      StreamObserver<InsertObjectRequest> requestStreamObserver =
+          stub.withDeadlineAfter(WRITE_STREAM_TIMEOUT.toMillis(), MILLISECONDS)
+              .insertObject(responseObserver);
+
+      boolean objectFinalized = false;
+      while (!objectFinalized) {
+        InsertObjectRequest insertRequest;
+        if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
+          insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
+          writeOffset += insertRequest.getChecksummedData().getContent().size();
+        } else {
+          ByteString data =
+              ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
+          dataChunkMap.put(writeOffset, data);
+          if (dataChunkMap.size() >= NUMBER_OF_REQUESTS_TO_RETAIN) {
+            dataChunkMap.remove(dataChunkMap.firstKey());
+          }
+          insertRequest = buildInsertRequest(writeOffset, data, false);
+          writeOffset += data.size();
         }
+        requestStreamObserver.onNext(insertRequest);
+        objectFinalized = insertRequest.getFinishWrite();
 
         if (responseObserver.hasTransientError()) {
-          writeOffset = getCommittedWriteSize(uploadId);
-          ++retriesAttempted;
-        } else {
-          writeOffset += responseObserver.bytesWritten();
-          retriesAttempted = 0;
+          retriesAttempted++;
+          return doResumableUpload();
         }
+      }
 
-        if (retriesAttempted >= UPLOAD_RETRIES) {
-          throw new IOException(
-              String.format(
-                  "Too many retry attempts. Resumable upload failed for '%s'", resourceId));
+      requestStreamObserver.onCompleted();
+
+      try {
+        responseObserver.done.await();
+        if (responseObserver.hasTransientError()) {
+          retriesAttempted++;
+          return doResumableUpload();
```

Comment:
This is users recursion for retries, can we use iteration instead - it should be much easier to follow iterative code.

Reply:
Updated, more so because of multiple places handling errors, recursion is actually more readable to me in that case. Ended up having to added nested while loops and more branches with iteration.